rfcs/0004-transaction.md

```diff
@@ -0,0 +1,258 @@
+
+## References
+
+- [Transactions - RocksDB](https://github.com/facebook/rocksdb/wiki/Transactions)
+- [Concurrent ACID Transactions in Badger](https://dgraph.io/blog/post/badger-txn/)
+- [Transaction Internals: RocksDB](https://flaneur2020.github.io/posts/2021-08-14-rocksdb-txn/)
+- [Transaction Internals: Badger](https://flaneur2020.github.io/posts/2021-08-01-badger-txn/)
+- [Serializable Snapshot Isolation in PostgreSQL](https://arxiv.org/pdf/1208.4179)
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn = kv_store.transaction();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Isolation Levels
+
+The Snapshot Isolation have a good reputation in the industry, and is widely used in many databases. It's gives a lightweight Snapshot on each transactions started, and checks whether the writes against the snapshot has modified by others or not during the commit. It is very efficient and is considered as a good default isolation level for most of the use cases.
+
+However, Snapshot Isolation still suffers Write Skew anomaly, for example:
+
+1. Initially, account A has $600, the account B has $500, and the account C has $0.
+2. Two transactions occur concurrently: Transaction 1 transfer $400 from account A to account C, while transaction 2 Transfer $300 from account A to account B.
+1. Transaction 1 commits: account A has $200, account B has $500, account C has $400.
+2. Transaction 2 commits: account A has $300, account B has $800, account C has $400.
+
+In this case, the total amount of money in the system has increased by $100, which is a violation of the integrity constraint that the total amount of money in the system should remain constant.
```

Comment:
This can't happen with snapshot isolation right, as both transactions would mutate account A and therefore conflict? I'm having trouble coming up with a practical example of write-skew. I guess there's the classic Transaction 1 reads A and writes its value into B, and Transaction 2 reads B and writes its value into A which can result in A and B being swapped, where in a serial execution you either have both A and B taking A's original value or B's original value.

Reply:
oh you're right, they both changed account A, so it's expected to have conflicts during commits. sorry for the misguiding, let me figure out another example on write skew...

Reply:
changed with another example in 793742

Reply:
changed with another example in [7937429](https://github.com/slatedb/slatedb/pull/260/commits/793742945ea43d3a82fc754bc2b89d753f36f4a6)